proto/asn1-pdu/asn1_universal_types_to_der.cc

```diff
@@ -0,0 +1,103 @@
+#include "asn1_universal_types_to_der.h"
+
+#include <google/protobuf/util/time_util.h>
+
+namespace asn1_universal_types {
+
+void Encode(const Integer& integer, std::vector<uint8_t>& der) {
+  // Save the current size in |tag_len_pos| to place tag and length
+  // after the value is encoded.
+  const size_t tag_len_pos = der.size();
+
+  if (!integer.val().empty()) {
+    der.insert(der.end(), integer.val().begin(), integer.val().end());
+  } else {
+    // Cannot have an empty integer, so use the value 0.
+    der.push_back(0x00);
+  }
+
+  EncodeTagAndLength(kAsn1Integer, der.size() - tag_len_pos, tag_len_pos, der);
+}
+
+void Encode(const BitString& bit_string, std::vector<uint8_t>& der) {
+  // Save the current size in |tag_len_pos| to place tag and length
+  // after the value is encoded.
+  const size_t tag_len_pos = der.size();
+
+  if (!bit_string.val().empty()) {
+    der.push_back(bit_string.unused_bits());
+    der.insert(der.end(), bit_string.val().begin(), bit_string.val().end());
+  } else {
+    // If the bitstring is empty, there shall be no subsequent octets,
+    // and the initial octet shall be zero (X.690 (2015), 8.6.2.3).
+    der.push_back(0x00);
+  }
+
+  EncodeTagAndLength(kAsn1Bitstring, der.size() - tag_len_pos, tag_len_pos,
+                     der);
+}
+
+void Encode(const UTCTime& utc_time, std::vector<uint8_t>& der) {
+  // Save the current size in |tag_len_pos| to place tag and length
+  // after the value is encoded.
+  const size_t tag_len_pos = der.size();
+
+  EncodeTimestamp(utc_time.time_stamp(), 13, der);
+
+  // Check if encoding was unsuccessful.
```

Comment:
does the condition in the `if` imply that the encoding was unsuccessful? Would it make sense to re-write it in the following way:

```cpp
  if (der.size() != tag_len_pos)
    EncodeTagAndLength(kAsn1UTCTime, der.size() - tag_len_pos, tag_len_pos, der);
```

and update the comment if needed

Reply:
same comment for another version of the method on line 55